### PR TITLE
[SPARK-42631][CONNECT] Support custom extensions in Scala client

### PR DESCRIPTION
**Pull request by tomvanbussel**
### What changes were proposed in this pull request?

This PR adds public interfaces for creating `Dataset` and `Column` instances, and for executing commands. These interfaces only allow creating `Plan`s and `Expression`s that contain an `extension` to limit what we need to expose.

### Why are the changes needed?

Required to implement extensions to the Scala Spark Connect client.

### Does this PR introduce _any_ user-facing change?

Yes, adds new public interfaces (see above).

### How was this patch tested?

Added unit tests.

### CODE REVIEW

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
The JIRA ID is wrong, isn't it? @tomvanbussel .

**Comment by tomvanbussel** [comment]
@dongjoon-hyun It is! Not sure how that happened... Should be fixed now!

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for updating.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala [hunk 1]
```diff
@@ -1312,6 +1312,10 @@ private[sql] object Column {
     new Column(builder.build())
   }
 
+  def apply(extension: com.google.protobuf.Any): Column = {
+    apply { builder => builder.setExtension(extension) }
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1316]: does this pass style checks?

OCD: this also works: `apply(_.setExtension(extension))`

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -120,7 +120,7 @@ import org.apache.spark.util.Utils
  */
 class Dataset[T] private[sql] (
     val sparkSession: SparkSession,
-    private[sql] val plan: proto.Plan,
+    val plan: proto.Plan,
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:123]: Can we mark this field as DeveloperApi. Not sure if we can do that at the field level though.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -261,6 +261,14 @@ class SparkSession private[sql] (
     new Dataset[T](this, plan, encoder)
   }
 
+  def newDataFrame(extension: com.google.protobuf.Any): DataFrame = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:264]: Please mark as `@DeveloperApi`.

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:264]: Same for all the other methods you have added here.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 4]
```diff
@@ -152,11 +161,9 @@ class PlanGenerationTestSuite
   }
 
   private def readRelation(path: Path): proto.Relation = {
-    val input = Files.newInputStream(path)
-    try proto.Relation.parseFrom(input)
-    finally {
-      input.close()
-    }
+    val builder = proto.Relation.newBuilder()
+    parser.merge(Files.readString(path), builder)
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:165]: Are you sure it is safe to read this as a string? The files we are reading are binary (so no valid UTF-8).

Comment (tomvanbussel) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala]: Oops, meant to revert this.

**Review by hvanhovell — APPROVED** [review]
LGTM

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (tomvanbussel) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala]: It's possible. It's also done for the non-Connect `Dataset`: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L191

**Comment by hvanhovell** [comment]
@dongjoon-hyun are you ok with this PR now?

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Sure! Thank you for checking again, @hvanhovell .

**Comment by hvanhovell** [comment]
Alright, merging this one.

**Pull request closed without merging**